### PR TITLE
feat: support using apisix client separately with cache bypass

### PR DESCRIPTION
**Pull request by Yiyiyimu**
Signed-off-by: yiyiyimu <wosoyoung@gmail.com>Please answer these questions before submitting a pull request

- Why submit this pull request?
- [ ] Bugfix
- [x] New feature provided
- [ ] Improve performance
- [ ] Backport patches

- Related issues

___
### New feature or improvement
- Describe the details and related test reports.

	- For using apisix go client separately without kubernetes environment, we need to add an option to bypass the cache.
	- Also add `GetById` for resources to directly get them with their real id.
___
### Backport patches
- Why need to backport?

- Source branch

- Related commits and pull requests

- Target branch

### CODE REVIEW

File: pkg/apisix/cluster.go [hunk 1]
```diff
@@ -276,6 +281,11 @@ func (c *cluster) String() string {
 
 // HasSynced implements Cluster.HasSynced method.
 func (c *cluster) HasSynced(ctx context.Context) error {
+	// if bypass cache no need to sync
+	if !c.bypassCache {
```
Comment (tokers) [inline pkg/apisix/cluster.go:285]: It should be `if c.bypassCache`.

File: pkg/apisix/route.go [hunk 2]
```diff
@@ -41,16 +41,24 @@ func newRouteClient(c *cluster) Route {
...
+}
+
 // FIXME, currently if caller pass a non-existent resource, the Get always passes
 // through cache.
-func (r *routeClient) Get(ctx context.Context, name string) (*v1.Route, error) {
+func (r *routeClient) get(ctx context.Context, id string, name string) (*v1.Route, error) {
 	log.Debugw("try to look up route",
 		zap.String("name", name),
 		zap.String("url", r.url),
 		zap.String("cluster", "default"),
```
Comment (tokers) [inline pkg/apisix/route.go:59]: Should also print the route id in logs.

File: pkg/apisix/route.go [hunk 3]
```diff
@@ -41,16 +41,24 @@ func newRouteClient(c *cluster) Route {
...
 	)
-	rid := id.GenID(name)
-	route, err := r.cluster.cache.GetRoute(rid)
+	route, err := r.cluster.cache.GetRoute(id)
```
Comment (tokers) [inline pkg/apisix/route.go:61]: It still queries the cache. Does this PR still work in progress?

File: pkg/apisix/cluster.go (hunk 1 above)
Comment (Yiyiyimu) [reply pkg/apisix/cluster.go:285]: Thanks! fixed

File: pkg/apisix/route.go (hunk 2 above)
Comment (Yiyiyimu) [reply pkg/apisix/route.go:59]: `GetById` also pass id as `name` in `get`. 
So changed `zap.String("name", name)` to `zap.String("name/id", name)`

File: pkg/apisix/route.go (hunk 3 above)
Comment (Yiyiyimu) [reply pkg/apisix/route.go]: Fixed. 
BTW, I noticed only consumers are not imported to cache. Is that on purpose?

Comment (tokers) [reply pkg/apisix/route.go]: Consumers will be saved into cache.

**Comment by gxthrj** [comment]
What is the usage scenario of this PR? Is there any related issue description? 
I am worried that the feature will cause the cache to be inconsistent with the internal state of APISIX.

**Review by tokers — COMMENTED** [review]
> What is the usage scenario of this PR? Is there any related issue description?
> I am worried that the feature will cause the cache to be inconsistent with the internal state of APISIX.

APISIX Ingress Controller won't use it, this is for other scenarios.

**Comment by Yiyiyimu** [comment]
> What is the usage scenario of this PR? Is there any related issue description?
> I am worried that the feature will cause the cache to be inconsistent with the internal state of APISIX.

Hi @gxthrj currently I'm using it as apisix go client in https://github.com/api7/kong-to-apisix

File: pkg/apisix/global_rule.go [hunk 4]
```diff
@@ -40,33 +40,46 @@ func newGlobalRuleClient(c *cluster) GlobalRule {
...
-			zap.Error(err),
-		)
-	} else {
-		log.Debugw("failed to find global_rule in cache, will try to lookup from APISIX",
-			zap.String("name", name),
-			zap.Error(err),
-		)
+
+	var globalRule *v1.GlobalRule
+	var err error
```
[the remark this answers is not in the document]
Comment (Yiyiyimu) [reply pkg/apisix/global_rule.go]: Thx! fixed

File: pkg/apisix/route.go (hunk 3 above)
Comment (Yiyiyimu) [reply pkg/apisix/route.go]: Got it. Will fixed it in another PR

File: pkg/apisix/route.go (hunk 2 above)
Comment (tokers) [reply pkg/apisix/route.go:59]: Here you don't change.

Comment (Yiyiyimu) [reply pkg/apisix/route.go:59]: 🤣 My bad. Fixed

**Review by tokers — APPROVED** [review]

**Review by tokers — CHANGES_REQUESTED** [review]
Won't merge as the E2E test was skipped unexpectedly.

**Comment by Yiyiyimu** [comment]
Close and would create another PR for new implementation

**Pull request closed without merging**